### PR TITLE
executor: attach to target cgroup earlier

### PR DESCRIPTION
**Pull request by bluca**
Try and reduce as much as possible the time during which the child is running in init.scope. The parent already moves the child after it's spawned too.

With the next glibc we will be able to spawn directly into the target cgroup and the problem will be solved comprehensively.

### CODE REVIEW

**Comment by mrc0mmand** [comment]
This unfortunately doesn't seem to help:

```
# uname -r
6.8.0-0.rc0.20240110gitab27740f7665.4.fc40.x86_64
# systemctl --version
systemd 255 (255-1091-g189fbc0)
# git log -2 --oneline
189fbc0201 (HEAD -> pr) Merge c9f420d8fd82d25343054881775c1cd54868c449 into 64b74c86d022eec9164f9c6ec80e16625b4ec27a
c9f420d8fd executor: attach to target cgroup earlier
# systemd-run -u foo.service --service-type=exec echo hello
Running as unit: foo.service; invocation ID: 128f0c721a8b40e2a362ee93e4b18490
# journalctl -o export -t echo
__CURSOR=s=53bcf061a49443dca354b4f433f18c83;i=a2a51;b=fb4ae0d506764842b90b366150da8312;m=111e4741;t=60ebe9b702638;x=653ff9e876f51ab
__REALTIME_TIMESTAMP=1705061674657336
__MONOTONIC_TIMESTAMP=287196993
__SEQNUM=666193
__SEQNUM_ID=53bcf061a49443dca354b4f433f18c83
_BOOT_ID=fb4ae0d506764842b90b366150da8312
SYSLOG_FACILITY=3
PRIORITY=6
_UID=0
_GID=0
_SELINUX_CONTEXT=kernel
_MACHINE_ID=1cc2313b3a074e32a4d61133e2586a5f
_HOSTNAME=localhost.localdomain
_RUNTIME_SCOPE=system
_CAP_EFFECTIVE=1ffffffffff
_SYSTEMD_CGROUP=/init.scope
_SYSTEMD_UNIT=init.scope
_SYSTEMD_SLICE=-.slice
_PID=4103
_COMM=16
_EXE=/usr/lib/systemd/systemd-executor
_CMDLINE=/usr/lib/systemd/systemd-executor --deserialize 128 --log-level debug --log-target journal-or-kmsg
_TRANSPORT=stdout
_STREAM_ID=911e76cd844b49f2becf19ec02e0f3c8
SYSLOG_IDENTIFIER=echo
MESSAGE=hello
```

**Comment by bluca** [comment]
Is it maybe because the logging is already set up by that time? Should it be shut down and reset after reattaching the cgroup?

File: src/core/exec-invoke.c [hunk 1]
```diff
@@ -4052,6 +4052,30 @@ int exec_invoke(
...
+                        *exit_status = EXIT_CGROUP;
+                        return log_exec_error_errno(context, params, r, "Failed to attach process to cgroup %s "
+                                                    "because the cgroup or one of its parents or "
+                                                    "siblings is in the threaded mode: %m", p);
+                }
+                if (r < 0) {
+                        *exit_status = EXIT_CGROUP;
+                        return log_exec_error_errno(context, params, r, "Failed to attach to cgroup %s: %m", p);
+                }
+        }
```
Comment (poettering) [inline src/core/exec-invoke.c:4081]: i think doing this probably makes sense (though it's not going to fix the issue, see other discussion), but I'd do it after the log_set_max_level() at least.

**Comment by mrc0mmand** [comment]
@bluca could you rebase & push? I hope the failing tests should addressed by https://github.com/systemd/systemd/pull/30999 and https://github.com/systemd/systemd/pull/30973.

**Comment by mrc0mmand** [comment]
With this change we apparently sometimes miss messages from journal namespaces, which makes `TEST-44-LOG-NAMESPACE` fail (it's easily reproducible by running the test a couple times):
```
[    4.712246] testsuite-44.sh[264]: + systemd-run --wait -p LogNamespace=foobar echo 'hello world'
[    4.943112] testsuite-44.sh[274]: Running as unit: run-u1.service; invocation ID: c325d7b6c7794e3dbdbf8aa376bdb59e
[    5.086555] testsuite-44.sh[274]: Finished with result: success
[    5.086555] testsuite-44.sh[274]: Main processes terminated with: code=exited/status=0
[    5.086555] testsuite-44.sh[274]: Service runtime: 33ms
[    5.086555] testsuite-44.sh[274]: CPU time consumed: 13ms
[    5.086555] testsuite-44.sh[274]: Memory peak: 804.0K (swap: 0B)
[    5.088084] testsuite-44.sh[264]: + systemd-run --wait -p LogNamespace=foobaz echo 'hello world'
[    5.103118] systemd-journald[279]: Failed to open /dev/kmsg, ignoring: Operation not permitted
[    5.326301] testsuite-44.sh[280]: Running as unit: run-u4.service; invocation ID: 7bc64b19424a424a9026eb701873c241
[    5.335168] testsuite-44.sh[280]: Finished with result: success
[    5.335168] testsuite-44.sh[280]: Main processes terminated with: code=exited/status=0
[    5.335168] testsuite-44.sh[280]: Service runtime: 41ms
[    5.335168] testsuite-44.sh[280]: CPU time consumed: 19ms
[    5.335168] testsuite-44.sh[280]: Memory peak: 624.0K (swap: 0B)
[    5.337936] testsuite-44.sh[264]: + journalctl --namespace=foobar --sync
[    5.350175] testsuite-44.sh[264]: + journalctl --namespace=foobaz --sync
[    5.360451] testsuite-44.sh[264]: + ls -l /var/log/journal/
[    5.373147] testsuite-44.sh[287]: total 16
[    5.373147] testsuite-44.sh[287]: drwxr-sr-x+ 2 root systemd-journal 4096 Jan 22 18:28 ff05055ea8a8441f8649851fc71f2ede
[    5.373147] testsuite-44.sh[287]: drwxr-sr-x+ 2 root systemd-journal 4096 Jan 22 18:28 ff05055ea8a8441f8649851fc71f2ede.foobar
[    5.374032] testsuite-44.sh[264]: + journalctl --list-namespaces
[    5.383512] testsuite-44.sh[288]: NAMESPACE
[    5.383512] testsuite-44.sh[288]: foobar
```

Here the `foobaz` journal namespace is not even created, since the message got lost along the way.

**Comment by mrc0mmand** [comment]
Hah, so it's apparently not caused by this PR, since I just saw it in https://github.com/systemd/systemd/pull/31064 as well...

**Comment by bluca** [comment]
Superseded by https://github.com/systemd/systemd/pull/31209

**Pull request closed without merging**